spec/src/main/asciidoc/property-expressions.asciidoc

```diff
@@ -0,0 +1,97 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// Contributors:
+// Roberto Cortez
+
+[[property-expressions]]
+== Property Expressions
+
+The value of a configuration property may contain an expression corresponding to another configuration property. An
+expression string is a mix of plain strings and expression segments, which are wrapped by the sequence `${ ... }`.
+
+Consider the following configuration properties file:
+
+[source,properties]
+----
+server.url=http://${server.host}/endpoint
+server.host=example.org
+----
+
+When looking up the `server.url` property, the value will be resolved and expanded to `http://example.org/endpoint`.
+All MicroProfile Config rules still apply. The `Config` is able to resolve expressions from different ConfigSources.
+
+Additionally, it is also possible to use the following syntax for property expressions:
+
+* `${expression:value}` - Provides a default value after the `:` if the expression doesn't find a value.
+* `${my.prop${compose}}` - Composed expressions. Inner expressions are resolved first.
+* `${my.prop}${my.prop}` - Multiple expressions.
+
+Consider the following configuration properties file:
+
+[source,properties]
+----
+server.url=http://${server.host:example.org}:${server.port}/${server.endpoint}
+server.port=8080
+server.endpoint=${server.endpoint.path.${server.endpoint.path.bar}}
+server.endpoint.path.foo=/foo
+server.endpoint.path.bar=foo
+----
+
+The property `server.url` is expanded to `http://example.org:8080/foo/bar`.
+
+If an expression cannot be expanded and does not have a default value, a `NoSuchElementException` is thrown. In the
+Optional case, an empty Optional will be returned.
+
+The number of recursion lookups is not infinite, but a limited number for composed expressions. Implementations are
+encouraged to limit the number to `5`, but they can use higher limits if they wish to. When the number of allowed
+lookups exceeds the limit, a `IllegalArgumentException` is thrown.
+
+Property expressions applies to all the methods in `Config` that performs resolution of a configuration property,
+including `getValue`, `getValues`, `getConfigValue`, `getValues`, `getOptionalValue`, `getOptionalValues`
+and `getConfigProperties`. The methods `getValue` and `getProperties` in `ConfigSource`, may support property
+expressions as well, but it is not required by the specification.
+
+Property expressions must also be applied to configuration properties injected via CDI. A default value
+defined via `org.eclipse.microprofile.config.inject.ConfigProperty#defaultValue` is not eligible to be expanded since
+multiple candidates may be available.
+
+If a configuration property value requires the raw value without expansion, the expression may be escaped with a
+double backslash ("\\"). For instance:
```

Comment:
Why double backslash? I recommend to use a single backslach to be consistent with the comma escape for array values.

Reply:
Well sure, is just that property file based sources requires a double backslash.